[WIP] Add PHP pkgconf .pc metadata file(s)

petk opened the pull request:
This is for the time being just a concept and work in progress so it can be properly synced with #12389 It will be rebased and changes gradually incorporated since it depends on other refactoring of the libdir variable name syncs and simplifications.

Since PHP *nix Autotools build system already extensively relies on [pkgconf (pkg-config)](https://people.freedesktop.org/~dbn/pkg-config-guide.html) command-line tool to find dependencies it should also provide its own .pc file(s).

This adds an initial php.pc.in and php-embed.pc.in template files that are created during the configuration and build phase. They are installed in the provided system pkgconfig directory. For example:

    /usr/lib/pkgconfig/php.pc
    /usr/lib/pkgconfig/php-embed.pc
    /usr/lib/x86_64-linux-gnu/pkgconfig/php.pc

[Pkgconf][1] is a more actively maintained continuation of the initial [Freedesktop's pkg-config project][2].

Extensions and applications embedding the PHP Embed SAPI can then optionally use also the pkgconf command-line tool to get the required PHP CFLAGS and/or LIBS.

    pkgconf --libs php-embed
    pkgconf --modversion php-embed
    pkgconf --cflags php-embed
    pkgconf --cflags php

[1]: https://github.com/pkgconf/pkgconf
[2]: https://gitlab.freedesktop.org/pkg-config/pkg-config

morrisonlevi commented:
I think we should add variables `php_debug` and `php_zts` as well.

petk commented:
> I think we should add variables `php_debug` and `php_zts` as well. Extensions need to know this information to order to know what ABIs to build for.

Yes, this list of variables needs to be expanded further. I've added `php_debug=yes|no` and `php_zts=yes|no` variables.

morrisonlevi commented:
I think it needs the `PHP_VERSION_ID` as well. The `php-config` calls this `--vernum`. I may try building against this today to see if I can get it to work.

petk commented:
> I think it needs the `PHP_VERSION_ID` as well. The `php-config` calls this `--vernum`. I may try building against this today to see if I can get it to work.

This one is added as `php_vernum`. `php_vernum=@PHP_VERSION_ID@`.

morrisonlevi commented:
I just built from source and this one didn't get expanded:

```
Cflags.private: @PHP_CFLAGS_PRIVATE@
```

Also, is it expected that it goes into `${prefix}/lib/php/pkgconf` instead of `${prefix}/lib/pkgconf`?

petk commented:
> I just built from source and this one didn't get expanded:
> 
> ```
> Cflags.private: @PHP_CFLAGS_PRIVATE@
> ```
> 
> Also, is it expected that it goes into `${prefix}/lib/php/pkgconf` instead of `${prefix}/lib/pkgconf`?

Fixed in the 3rd commit. Installation now puts it to pkgconf directory in libdir. For now, I've removed this Cflags.private part.

Girgias commented:
@morrisonlevi could you have another look to see if this behaves like you expect it?

I have no knowledge about build systems/package dependency resolution for system libraries, but I think or we be good to use what seems to be industry standard.

petk commented:
I've rebased the branch so it's easier to handle this PR. Here actually everything works but PHP has a bit of weird include flags handling at the moment. This pkg-config template files are (for now) done like the php-config script is done. I'll recheck all the pkg-config variables again what needs to be added.

As far as the PHP-8.4 is concerned I won't touch the include flags yet in this branch. Because they aren't done correctly. Neither the main PHP include directory is done correctly. This also goes for the php-config script. They are for now hard-coded in there as part of the PHP history, but they would need to be dynamic.

deltragon commented:
> Since PHP *nix Autotools build system already extensively relies on pkgconf (pkg-config) command-line tool to find dependencies it should also provide its .pc metadata file(s). [1](https://people.freedesktop.org/~dbn/pkg-config-guide.html)
> 
> This adds an initial php.pc.in and php-embed.pc.in templates that are created during the configuration and build phase. They are installed in the provided system pkgconfig directory. For example:
> 
> ```
> /usr/lib/pkgconfig/php.pc
> /usr/lib/pkgconfig/php-embed.pc
> /usr/lib/pkgconfig/php-embed8.pc
> /usr/lib/pkgconfig/php-embed8.4.pc
> /usr/lib/x86_64-linux-gnu/pkgconfig/php.pc
> ```
> 
> Pkgconf [2](https://github.com/pkgconf/pkgconf) is a maintained continuation of the initial Freedesktop's pkg-config project [3](https://gitlab.freedesktop.org/pkg-config/pkg-config).
> 
> Extensions and applications embedding the PHP Embed SAPI can then use also pkgconf to get the required PHP CFLAGS and/or LIBS.
> 
> For PHP extensions:
> 
> ```
> pkgconf --cflags php
> pkgconf --libs php
> pkgconf --modversion php
> pkgconf --cflags php8.4
> ```
> 
> Applications embedding the PHP Embed SAPI:
> 
> ```
> pkgconf --libs php-embed
> pkgconf --cflags php-embed
> pkgconf --modversion php-embed
> pkgconf --cflags php-embed8.4
> ```

petk commented:
I've also added a quick table for tracking options between php-config script and possible pkg-config alternatives in the PR's first post. The next appended commit adds some fixes and syncs of the .pc.in templates.

petk commented:
Yes, this pkg-config stuff is complex, anyone here to test it :D?

Basically it works but the main issue with this will be in scenarios where different PHP versions are installed (where also php-config is prone to errors) or when installing PHP with program prefix/sufix (like php8.4), or when in extreme case both static and shared libphp embed is installed. The PHP Embed library still probably needs one more adjustment to totally match the php-config's `php-config --lib-embed` behavior. But I think we can adjust these .pc files on the go as needed, either through PHP 8.4 bug fixes or add more fine-tuned features in the next version...

petk commented:
I got another idea about the `php_zts` and `php_debug` variables. Instead of these only a single pkg-config variable like `php_features` can be added:

```diff
--- a/scripts/php.pc.in
+++ b/scripts/php.pc.in
@@ -12,10 +12,7 @@ php_vernum=@PHP_VERSION_ID@
-# Whether PHP is built in debug mode (yes) or not (no).
-php_debug=@ZEND_DEBUG@
-# Whether PHP is built with thread safety (yes) or not (no).
-php_zts=@PHP_THREAD_SAFETY@
+php_features="@PHP_FEATURES@"
```

This php_features would be a space separated list of enabled features, for example:

```
php_features="DEBUG ZTS"
```

Then the `php_features` can be parsed manually by the build system where needed. And it can be a bit simpler extended in the future with more "features" if needed. For example, curl is using a similar approach and I think it works pretty ok.

Anyhow, I think that this won't be able to land in PHP 8.4 so, let's target next PHP version here then.

cmb69 commented on CONTRIBUTING.md:
```diff
@@ -235,8 +235,11 @@ locations.
+       └─ ...
     └─ ...
- ├─ scripts/                        # php-config, phpize and internal development scripts
+ ├─ scripts/                        # php-config, phpize, pkg-config, and internal development scripts
```
This might better be `pkg-config metadata file` or maybe just `.pc`, or something like this.

cmb69 commented on configure.ac:
```diff
@@ -1467,7 +1469,7 @@ PHP_SUBST_OLD([EXTENSION_DIR])
 PHP_SUBST([EXTRA_LDFLAGS])
 PHP_SUBST([EXTRA_LDFLAGS_PROGRAM])
 PHP_SUBST_OLD([EXTRA_LIBS])
-PHP_SUBST([ZEND_EXTRA_LIBS])
+PHP_SUBST_OLD([ZEND_EXTRA_LIBS])
```
Out of curiosity, what's the difference between these macros?

cmb69 commented on scripts/php.pc.in:
```diff
@@ -0,0 +1,27 @@
+php_zts=@PHP_THREAD_SAFETY@
+
+Name: PHP
+Description: Build extension for a PHP general-purpose scripting language
```
Hmm, isn't this supposed to be the description of the package? So maybe instead
```suggestion
Description: A popular general-purpose scripting language that is especially suited to web development
```

petk replied:
Yes, probably that sounds a bit better. Issue is that with pc files there are two main workflows:
- when building PHP extension
- when building application that uses libphp embed SAPI (php-embed.pc)

With this description, I was aiming at making it more clear why the php.pc is actually used for. It wouldn't be used for anything besides building PHP extensions, from my current understandings.

cmb69 commented on scripts/php.pc.in:
```diff
@@ -0,0 +1,27 @@
+License: PHP
+Version: @PHP_VERSION@
+Cflags: -I${includedir} -I${includedir}/main -I${includedir}/TSRM -I${includedir}/Zend -I${includedir}/ext -I${includedir}/ext/date/lib
+Libs.private: @EXTRA_LIBS@ @ZEND_EXTRA_LIBS@
```
Do we need to add `@PHP_LDFLAGS@` here?

cmb69 reviewed (commented):
Generally, this looks like a very good (and overdue) addition (we could get rid of php-config in the long run). Thank you!

> I got another idea about the `php_zts` and `php_debug` variables. Instead of these only a single pkg-config variable like `php_features` can be added:

I think this is a good idea. (I would probably go with `features` only, though.)

petk commented:
I'm just updating current status here a bit. `.pc` files are very difficult to properly construct from the build system point of view for complex projects such as PHP.

There is also [CMake-based PHP build system](https://github.com/petk/php-build-system/) being developed, which does provide these `.pc` files as an option to find PHP on the host system. It is currently primarily using the CMake config files as a better alternative to just pc files. And on top of that, C/C++ community (and mainly CMake) are currently pushing the [CPS](https://cps-org.github.io/cps/) as a better alternative to pkg-config, which might provide a more useful alternative to pkg-config in the future.

PHP extensions are so far being able to find PHP on the host system through phpize and PIE tool, so pkg-config isn't that needed at this point I believe. I'll check more later what to do with pc files and this PR.

P.S.: BTW, thanks to @morrisonlevi and @cmb69 for taking time to review this PR, I appreciate it.
